Implemented Ticklabel selection when formatting - #8128

mitko1994 opened the pull request:
This is simple feature implementation which allows the user to select which tick labels to format when using figure.Figure. autofmt_xdate() - see #8128

anntzer commented on lib/matplotlib/figure.py:
```diff
@@ -440,38 +440,48 @@ def set_tight_layout(self, tight):
         self.stale = True
 
-    def autofmt_xdate(self, bottom=0.2, rotation=30, ha='right'):
+    def autofmt_xdate(self, bottom=0.2, rotation=30, ha='right',which=None):
```
The pep8 violations, here and below, will cause the test suite to fail.

anntzer commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -4972,3 +4972,32 @@ def test_bar_single_height():
+	    #check if the major axis were formatted (or not) as expected
+        for label in fig.axes[0].get_xticklabels(False,'major'):
+            print(label.get_rotation())
+            assert (label.get_rotation() == 30) == is_rotated_major
```
```
if is_rotated_major:
    assert label.get_rotation() == 30
else:
    ...
```
and you won't need to print the value of `label.get_rotation()`.

mitko1994 replied:
Oops! That print was only meant to be there for debugging when I was running the test on my own. Would removing it be good enough? The test for the minor labels below has the same format and I guess it's acceptable, right?

anntzer replied:
just remove it then.

mitko1994 replied:
The requested changes are now implemented.

mitko1994 commented:
This is my first ever pull request so I apologize for the rookie mistakes I made. I fixed the pep8 errors and removed the print statement from the test. I assume I don't have to submit a new PR with the changes, right?

anntzer commented:
No problem, this PR remains valid.

QuLogic commented:
Is that really the email address you want to use for attributing your commits? (It's currently the username@<virtual machine host name>) instead of a real address.)

mitko1994 commented:
Sorry, I didn't realize my email wasn't set for git. I set it now. I'm not sure it will reflect in previous commits though. Now that I fixed the  requested changes, what is the procedure for getting this PR merged?

mitko1994 commented:
I'm not sure what's wrong with the Travis CI build. I imported pandas the same way it's done in other tests. Can someone give me a suggestion on what I'm doing wrong?

QuLogic commented:
No, your previous commits will not use the correct address unless you modify them yourself. I can do that for you, but you will need to reset your local copy. Also, in the future, please rebase instead of merging master.

The test is not running into problems importing Pandas, it is failing to plot anything at all. Did you try running the test locally?

mitko1994 commented:
Thanks for your reply.  Email address: dimitar.stratiev[at]mail.utoronto.ca. I ran the test but not using pytest. I ran the method to make sure the functionality is as expected.

QuLogic commented:
I fixed the author information, but then I decided to squash everything. You will need to update your local copy. Assuming `upstream` is this repo and `origin` is your fork and you have a local branch of the same name:
```bash
$ git checkout feature#8128
$ git status
# Verify there are no changes you want to save; checkout or stash them first.

$ git fetch origin
$ git reset --hard origin/feature#8128

# Restore any changes or unstash them if there were any.
```

In the future, please rebase instead of merging `master`.

QuLogic commented:
I did not fix the tests; you will have to see about that yourself. It's `pytest /path/to/file.py`, not `pytest tests.py` (which contains no tests). So to check the file you change, use `pytest lib/matplotlib/tests/test_axes.py`.

mitko1994 commented:
@QuLogic I'm not exactly sure what's going on but when I run `pytest lib/matplotlib/tests/test_axes.py` I get a different error than that given in Travis CI. The error I get is as follows:

`dimitar@dimitar-VirtualBox:~/Desktop/matplotlib$ pytest lib/matplotlib/tests/test_axes.py
===========================  test_axes.py  ===========================
/usr/lib/python2.7/dist-packages/pkg_resources.py:1031: UserWarning: /home/dimitar/.python-eggs is writable by group/others and vulnerable to attack when used with get_resource_filename. Consider a more secure location (set with .set_extraction_path or the PYTHON_EGG_CACHE environment variable).
  warnings.warn(msg, UserWarning)
/usr/local/lib/python2.7/dist-packages/matplotlib-2.0.0_3568.g5c6b6d9-py2.7-linux-i686.egg/matplotlib/testing/decorators.py:453: UserWarning: Module 'test_axes' does not live in a parent module named 'tests'. This is probably ok, but we may not be able to guess the correct subdirectory containing the baseline images. If things go wrong please make sure that there is a parent directory named 'tests' and that it contains a __init__.py file (can be empty).
  "__init__.py file (can be empty).") % module_name)
unhandled exception occurred while testing test_axes
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/logilab/common/pytest.py", line 428, in testfile
    options=self.options, outstream=sys.stderr)
  File "/usr/lib/python2.7/dist-packages/logilab/common/pytest.py", line 764, in __init__
    testLoader=NonStrictTestLoader())
  File "/usr/lib/python2.7/unittest/main.py", line 77, in __init__
    self.module = __import__(module)
  File "test_axes.py", line 41, in <module>
    'formatter_ticker_005',
  File "/usr/local/lib/python2.7/dist-packages/matplotlib-2.0.0_3568.g5c6b6d9-py2.7-linux-i686.egg/matplotlib/testing/decorators.py", line 350, in __call__
    self.delayed_init(func)
  File "/usr/local/lib/python2.7/dist-packages/matplotlib-2.0.0_3568.g5c6b6d9-py2.7-linux-i686.egg/matplotlib/testing/decorators.py", line 250, in delayed_init
    self.baseline_dir, self.result_dir = _image_directories(func)
  File "/usr/local/lib/python2.7/dist-packages/matplotlib-2.0.0_3568.g5c6b6d9-py2.7-linux-i686.egg/matplotlib/testing/decorators.py", line 454, in _image_directories
    subdir = os.path.join(*mods)
TypeError: join() takes at least 1 argument (0 given)
*******************************************************************************
Ran 1 test cases in 0.00s (0.00s CPU), 1 errors
0 modules OK (1 failed)
failures: lib/matplotlib/tests/test_axes [1/1]`

QuLogic commented:
@mitko1994 It appears you have the logilab command `pytest`, which is _not_ the testing framework used by Matplotlib. Maybe try `py.test` instead?

mitko1994 commented:
I get yet another error that's different. This time it cannot import "ft2font".

stratiev@cmsvm24:~$ py.test matplotlib/lib/matplotlib/tests/test_axes.py
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/main.py", line 77, in wrap_session
INTERNALERROR>     config.do_configure()
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/config.py", line 607, in do_configure
INTERNALERROR>     self.hook.pytest_configure(config=self)
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/core.py", line 376, in __call__
INTERNALERROR>     return self._docall(methods, kwargs)
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/core.py", line 387, in _docall
INTERNALERROR>     res = mc.execute()
INTERNALERROR>   File "/usr/lib/python2.7/dist-packages/_pytest/core.py", line 288, in execute
INTERNALERROR>     res = method(**kwargs)
INTERNALERROR>   File "/home/stratiev/matplotlib/lib/matplotlib/testing/conftest.py", line 12, in pytest_configure
INTERNALERROR>     matplotlib._init_tests()
INTERNALERROR>   File "/home/stratiev/matplotlib/lib/matplotlib/__init__.py", line 1475, in _init_tests
INTERNALERROR>     from matplotlib import ft2font
INTERNALERROR> ImportError: cannot import name ft2font

mitko1994 commented:
Can someone please help me with this? I can't seem to be able to run pytest locally.

QuLogic commented:
Did you install Matplotlib in an environment or globally? Where did you install pytest (appears to be globally)? Make sure you install them both into the same place, preferably a virtualenv or conda environment. It's also preferable to install Matplotlib in development mode.

See the [development documentation for details](http://matplotlib.org/devdocs/devel/index.html).

anntzer commented:
Superseded by #8157.
